src/hooks/usePromiseDialog/PromiseDialogProvider.tsx

```diff
@@ -0,0 +1,85 @@
+import React from 'react';
+
+import omit from 'lodash/omit';
+
+import {PromiseDialogContext} from './PromiseDialogContext';
+import type {DialogRendererProps, PromiseDialogResult} from './types';
+
+type PromiseDialogProviderProps = {
+    children: React.ReactNode | React.ReactNode[];
+};
+
+export const PromiseDialogProvider = ({children}: PromiseDialogProviderProps) => {
+    const [dialogs, setDialogs] = React.useState<Record<number, React.ReactNode>>([]);
+    const dialogsRef: React.MutableRefObject<Record<number, React.ReactNode>> =
+        React.useRef(dialogs);
+
+    React.useEffect(() => {
+        dialogsRef.current = dialogs;
+    }, [dialogs]);
+
+    const contextValue = React.useMemo(
+        () => ({
+            openDialog: <ResultType extends unknown>(
+                renderDialog: ({
+                    onSuccess,
+                    asyncOnSuccess,
+                    onCancel,
+                    key,
+                }: DialogRendererProps<ResultType>) => React.ReactNode,
+            ) =>
+                new Promise<{success: boolean; value?: ResultType}>((resolve) => {
+                    const currentKeys = Object.keys(dialogsRef.current);
+
+                    const key = parseInt(currentKeys[currentKeys.length - 1] || '0', 10) + 1;
+
+                    const handleClose = (result: PromiseDialogResult<ResultType>) => {
+                        resolve(result);
+
+                        setDialogs(omit(dialogsRef.current, key));
+                    };
+
+                    const handleSuccess = (value: ResultType) => {
+                        handleClose({success: true, value});
+                    };
+
+                    const handleSuccessPromise = (
+                        getValue: Promise<ResultType>,
+                        onError: (error: unknown) => void,
```

Comment:
let's make this `onError` optional and make it possible to process the promise of the dialog using try catch?
```tsx
  getValue: Promise<ResultType>,
 onError: ((error: unknown) => void) | null,

try {
  const result = await useDialog(...)
} catch (e) {

}

```ts

pass null **explicitly** so as not to forget to handle the error in another way

Reply:
The hook returns an object with { success: boolean, value }, so there is no need to add the catch block
I think that onError must be required, because the user should handle their promise error to avoid unexpected behaviour like the form was sent, but nothing happened